File: MWSE/TES3Enchantment.cpp
#include "TES3Enchantment.h"

namespace TES3 {
	const auto TES3_Enchantment_ctor = reinterpret_cast< void( __thiscall * )( Enchantment * ) >( 0x4AADA0 );
	Enchantment::Enchantment() :
		objectID{},
		castType{},
		chargeCost{},
		maxCharge{},
		pad_32{},
		effects{},
		flags{}
	{
		TES3_Enchantment_ctor( this );
	}

	const auto TES3_Enchantment_dtor = reinterpret_cast< void( __thiscall * )( Enchantment * ) >( 0x4AAEA0 );
	Enchantment::~Enchantment()
	{
		TES3_Enchantment_dtor( this );
	}

	size_t Enchantment::getActiveEffectCount() {
		size_t count = 0;
		for (size_t i = 0; i < 8; i++) {
			if (effects[i].effectID != TES3::EffectID::None) {
				count++;
			}
		}
		return count;
	}

	int Enchantment::getFirstIndexOfEffect(int effectId) {
		for (size_t i = 0; i < 8; i++) {
			if (effects[i].effectID == effectId) {
				return i;
			}
		}
		return -1;
	}

	std::reference_wrapper<Effect[8]> Enchantment::getEffects() {
		return std::ref(effects);
	}
}

MWSE_SOL_CUSTOMIZED_PUSHER_DEFINE_TES3(TES3::Enchantment)

File: misc/package/Data Files/MWSE/core/meta/class/tes3enchantment.lua
--- @meta

--- An enchantment game object.
--- @class tes3enchantment : tes3object, tes3baseObject
--- @field castType number The enchantment's cast type. Maps to tes3.enchantmentType.* constants.
--- @field chargeCost number The cost of using the enchantment.
--- @field effects table *Read-only*. An array-style table of the tes3effect data on the object.
--- @field flags number A bit field for the enchantment's flags.
--- @field maxCharge number The maximum charge for the associated enchantment.
tes3enchantment = {}

--- Use `tes3.createObject` instead. This function created a new enchantment object, which will be stored as part of the current saved game.
--- @deprecated
--- @param params tes3enchantment.create.params This table accepts the following values:
--- 
--- `id`: string — *Optional*. The new object's ID. Must be unique if provided.
--- 
--- `castType`: number — The enchantment castType. See tes3.enchantmentType.
--- 
--- `chargeCost`: number — The new enchantment charge cost. Must be greater than 0.
--- 
--- `maxCharge`: number — The new enchantment maximum charge. Must be greater than 0
--- 
--- `flags`: number — *Optional*. The new enchantment flags.
--- 
--- `objectFlags`: number — *Default*: `0`. The object flags initially set. Force set as modified.
function tes3enchantment.create(params) end

---Table parameter definitions for `tes3enchantment.create`.
--- @class tes3enchantment.create.params
--- @field id string *Optional*. The new object's ID. Must be unique if provided.
--- @field castType number The enchantment castType. See tes3.enchantmentType.
--- @field chargeCost number The new enchantment charge cost. Must be greater than 0.
--- @field maxCharge number The new enchantment maximum charge. Must be greater than 0
--- @field flags number *Optional*. The new enchantment flags.
--- @field objectFlags number *Default*: `0`. The object flags initially set. Force set as modified.

--- Returns the amount of effects the tes3enchantment object has.
--- @return number count No description yet available.
function tes3enchantment:getActiveEffectCount() end

--- Returns the index of a first effect of a given effectId in the parent tes3enchantment object.
--- @param effectId number The effectID to perform a check for. Maps to tes3.effect.* constants.
--- @return number index No description yet available.
function tes3enchantment:getFirstIndexOfEffect(effectId) end

